### PR TITLE
Add account history

### PR DESCRIPTION
**Pull request by CristianFodor**
## Type
- [ ] Bug
- [x] Feature  
- [ ] Performance improvement

## Problem setting
Extracting data regarding account balance history and account token balance history
  
## Proposed Changes
- 
- 
- 

## How to test
devnet-api.elrond.com/accounts/:address/history
devnet-api.elrond.com/accounts/:address/history/:tokenIdentifier

### CODE REVIEW

**Comment by AlexandruPislariu** [comment]
Hello Cristian! Thank you for your contribution.
Can you please solve the conflicts in your local fork?

**Comment by CristianFodor** [comment]
> Hello Cristian! Thank you for your contribution. Can you please solve the conflicts in your local fork?

Hi Alexandru! I did it, please have a look again.

**Review by AlexandruPislariu — COMMENTED** [review]
Awesome integration! Thank you for contributing to our public API.
I have just a few small comments.

File: src/endpoints/accounts/account.service.ts [hunk 1]
```diff
@@ -345,4 +348,46 @@ export class AccountService {
...
+    elasticQuery
+        .withPagination(pagination)
+        .withSort([{ name: 'timestamp', order: ElasticSortOrder.descending }]);
+
+    const elasticResult = await this.elasticService.getList('accountshistory', 'address', elasticQuery);
+    return elasticResult.map(item => ApiUtils.mergeObjects(new AccountHistory(), item));
+  }
+
+  private static buildAccountHistoryFilterQuery(address?: string, token?: string): ElasticQuery {
+    const shouldQueries: AbstractQuery[] = [];
```
Comment (AlexandruPislariu) [inline src/endpoints/accounts/account.service.ts:364]: Can be removed since it isn't used

File: src/endpoints/accounts/account.service.ts [hunk 2]
```diff
@@ -345,4 +348,46 @@ export class AccountService {
...
+      mustQueries.push(QueryType.Match('token', token));
+    }
+
+    return ElasticQuery.create()
+      .withCondition(QueryConditionOptions.should, shouldQueries)
+      .withCondition(QueryConditionOptions.must, mustQueries);
+  }
+
+  async getAccountTokenHistory(address: string, tokenIdentifier: string, pagination: QueryPagination): Promise<AccountEsdtHistory[]> {
+    if(!tokenIdentifier){
```
Comment (AlexandruPislariu) [inline src/endpoints/accounts/account.service.ts:381]: The tokenIdentifier can't be undefined or null so this check will have no effect.

File: src/endpoints/accounts/account.controller.ts [hunk 3]
```diff
@@ -769,4 +771,45 @@ export class AccountController {
...
     return scResult;
   }
+
+  @ApiQuery({ name: 'from', description: 'Number of items to skip for the result set', required: false })
+  @ApiQuery({ name: 'size', description: 'Number of items to retrieve', required: false })
+  @Get("/accounts/:address/history")
+  @ApiResponse({
+    status: 200,
+    description: 'The account EGLD balance historical data for given address',
+    type: SmartContractResult,
```
Comment (AlexandruPislariu) [inline src/endpoints/accounts/account.controller.ts:781]: Here the type should be AccountHistory[]

File: src/endpoints/accounts/account.controller.ts [hunk 4]
```diff
@@ -769,4 +771,45 @@ export class AccountController {
...
+    return this.accountService.getAccountHistory(address, { from, size });
+  }
+
+  @ApiQuery({ name: 'from', description: 'Number of items to skip for the result set', required: false })
+  @ApiQuery({ name: 'size', description: 'Number of items to retrieve', required: false })
+  @Get("/accounts/:address/history/:tokenIdentifier")
+  @ApiResponse({
+    status: 200,
+    description: 'The token balance history for given address',
+    type: SmartContractResult,
```
Comment (AlexandruPislariu) [inline src/endpoints/accounts/account.controller.ts:801]: Also, here the type should be AccountEsdtHistory[]

File: src/endpoints/accounts/entities/account.esdt.history.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { ApiProperty } from "@nestjs/swagger";
+
+export class AccountEsdtHistory {
```
Comment (AlexandruPislariu) [inline src/endpoints/accounts/entities/account.esdt.history.ts:3]: I think this class should be inherited from AccountHistory and just add the token property in addition.

**Review by AlexandruPislariu — CHANGES_REQUESTED** [review]

**Review by AlexandruPislariu — APPROVED** [review]

**Review by tanghel — APPROVED** [review]

**Pull request merged**